ksqldb-engine/src/main/java/io/confluent/ksql/services/DefaultConnectClientFactory.java

```diff
@@ -0,0 +1,150 @@
+        // if no explicit header specified, then forward incoming request header
+        connectAuthHeader.isPresent() ? connectAuthHeader : ksqlAuthHeader
+    );
+  }
+
+  @Override
+  public synchronized void close() {
+    if (credentialsFileWatcher != null) {
+      credentialsFileWatcher.shutdown();
+    }
+  }
+
+  private Optional<String> buildAuthHeader() {
+    // custom basic auth credentials
+    if (ksqlConfig.getString(KsqlConfig.CONNECT_BASIC_AUTH_CREDENTIALS_SOURCE_PROPERTY)
+        .equalsIgnoreCase(KsqlConfig.BASIC_AUTH_CREDENTIALS_SOURCE_FILE)) {
+      final String credentialsFile =
+          ksqlConfig.getString(KsqlConfig.CONNECT_BASIC_AUTH_CREDENTIALS_FILE_PROPERTY);
+      final boolean failOnUnreadableCreds =
+          ksqlConfig.getBoolean(KsqlConfig.CONNECT_BASIC_AUTH_FAIL_ON_UNREADABLE_CREDENTIALS);
+
+      if (ksqlConfig.getBoolean(KsqlConfig.CONNECT_BASIC_AUTH_CREDENTIALS_RELOAD_PROPERTY)) {
+        startBasicAuthFileWatcher(credentialsFile, failOnUnreadableCreds);
+      }
+
+      return buildBasicAuthHeader(credentialsFile, failOnUnreadableCreds);
+    } else {
+      return Optional.empty();
+    }
+  }
+
+  private void startBasicAuthFileWatcher(
+      final String filePath,
+      final boolean failOnUnreadableCreds
+  ) {
+    try {
+      credentialsFileWatcher = new FileWatcher(Paths.get(filePath), () -> {
+        connectAuthHeader = buildBasicAuthHeader(filePath, failOnUnreadableCreds);
+      });
+      credentialsFileWatcher.start();
+      log.info("Enabled automatic connector credentials reload for location: " + filePath);
+    } catch (java.io.IOException e) {
+      log.error("Failed to enable automatic connector credentials reload.");
+    }
+  }
+
+  private static Optional<String> buildBasicAuthHeader(
+      final String credentialsPath,
+      final boolean failOnUnreadableCredentials
+  ) {
+    if (credentialsPath == null || credentialsPath.isEmpty()) {
+      throw new ConfigException(String.format("'%s' cannot be empty if '%s' is set to '%s'",
+          KsqlConfig.CONNECT_BASIC_AUTH_CREDENTIALS_FILE_PROPERTY,
+          KsqlConfig.CONNECT_BASIC_AUTH_CREDENTIALS_SOURCE_PROPERTY,
+          KsqlConfig.BASIC_AUTH_CREDENTIALS_SOURCE_FILE));
+    }
+
+    final Properties credentials = new Properties();
+    FileInputStream inputStream = null;
+    try {
+      inputStream = new FileInputStream(credentialsPath);
+      credentials.load(inputStream);
+
+      if (credentials.containsKey(KsqlConfig.BASIC_AUTH_CREDENTIALS_USERNAME)
+          && credentials.containsKey(KsqlConfig.BASIC_AUTH_CREDENTIALS_PASSWORD)) {
+        final String userInfo = credentials.getProperty(KsqlConfig.BASIC_AUTH_CREDENTIALS_USERNAME)
+            + ":" + credentials.getProperty(KsqlConfig.BASIC_AUTH_CREDENTIALS_PASSWORD);
+        return Optional.of("Basic " + Base64.getEncoder()
+            .encodeToString(userInfo.getBytes(Charset.defaultCharset())));
```

Comment:
Seems we need a common class to encode user/pass. This is used in `JaasPrincipal` and also here. 
i.e.
```
// Returns a base64 encode user/password
String KsqlCredentials.basic(user, pass);   
```

Reply:
Yeah, it's used all over the place. Also in RestIntegrationTestUtil.java and ClientImpl.java. Probably other places too.

It's a small enough piece of code and it doesn't look like there's an existing util file that makes sense for a helper method; that's probably why we haven't created one yet. Do you think it's worth adding a new file such as BasicAuthUtil into ksqldb-common just for this? My instinct says no but I can add one if you think it's better.